### ISSUE TITLE
SFTP: Corrupt files after download

### ISSUE THREAD

**Issue by hamza-am100** [phpseclib/phpseclib#1343]
Downloading files via SFTP end up in corrupted files. When i request the size using the function `$sftp-size($file)` the file size is correct, but after downloading the size of the result is always different. It does not matter what the size of the file is that i'm trying to download and downloading files in chunks also result in a partial files.

The server uses `SFTP protocol version 3`, the log files can be found in the following links.
ssh-log: https://drive.google.com/open?id=1NOg1E1Iu_LoraVhHqUw0DsRU8DJn1XJK
error-log: https://drive.google.com/open?id=1hIEyZfd3fRtzRuJaIFKfD-SwC9SFamkw

## Example
```
#1 (logs from this scenario provided)
File size: 552078
Result size: 388238
Result using chunk: 491520

#2
File size: 52646075
Result size: 52428800
Result using chunk: 52428800

#3
Full size: 100000132
Result size: 99614720,
Result using chunk: 99614720
```

Thanks in advanced.

**Comment by terrafrost** [comment]
I tried to access the files on the drive.google.com links you provided and I guess I need you to provide me access. The email address I tried to access them with is terrafrost@gmail.com...

**Comment by hamza-am100** [comment]
I've added you to the list.

**Comment by terrafrost** [comment]
You shared error-log with me. Could you share ssh-log with me as well?

Is ssh-log the OpenSSH log or the phpseclib logs? I would much prefer the phpseclib logs.

Thanks!

**Comment by hamza-am100** [comment]
I shared both ssh-logs and ssh-error-logs. The ssh-logs are obtained using the phplibsec functions
```
define('NET_SFTP_LOGGING', NET_SFTP_LOG_COMPLEX);
define('NET_SSH2_LOGGING', NET_SSH2_LOG_COMPLEX);

// initializition and downlowing file ......

file_put_contents("sftp-errors.log", $sftp->getErrors());
file_put_contents("sftp.log", $sftp->getLog());
```

**Comment by hamza-am100** [comment]
Could it be related to the server? Its a Amazon webserver using linux.

**Comment by terrafrost** [comment]
I hadn't had a chance to start going through the log files until today (busy past few days at work). My work is ongoing.

My initial theory (with the caveat that I'm not finished going through the log files) is that you're downloading a file that's actively being modified (it kinda looks like you're downloading Apache server logs? keep in mind I'm still going through stuff).

A good example of this would be...  let's say a file has 1mb added to it every 30s. I start the download when it's at 5mb and the download takes 1m. That'd mean that between the time the download started and ended the file grew in size by 2mb to 7mb. In that scenario phpseclib would only download the first 5mb since the file size was 5mb when it started.

Again, this is based on my *very* cursory glance of what I've seen thus far and my assessment could change the more I delve into what you've provided.

**Comment by terrafrost** [comment]
Here's the packet that's returning the size of the file from the extracted SFTP logs:

```
<- NET_SFTP_ATTRS
00000000  00:00:00:0d:00:00:00:00:00:08:6c:8e:00:00:81:e4  ..........l.....
00000010  5c:38:b4:a4:5c:38:b4:a4                          \8..\8..
```

Here's a breakdown of the first 12 bytes in that packet:

- **00:00:00:0d**: flags. the following are present:
  SSH_FILEXFER_ATTR_SIZE
  SSH_FILEXFER_PERMISSIONS
  SSH_FILEXFER_ATTR_ACMODTIME

- **00:00:00:00:00:08:6c:8e**: size (552078 bytes)

I extracted the file (export_2018-12-20.csv) from the logs and it is the exact length that the NET_SFTP_ATTRS packet claimed it was. You say that the result size is 388238 and that the result using chunk is 491520. How are "result size" and "result using chunk" different? Can you post the code you're using to do the download? I can email you the file I was able to extract from the logs if you want to compare it to the expected file.

**Comment by hamza-am100** [comment]
Thank you for your time.

To reply to your previous message, i don't think this is related to files being updated at the time of downloading the file. This is a structural problem, it occurs with every download and independent of the file.


I don't run out of memory, i also used an implementation where i store the result directly to a file. This is the code that i use for downloading the files.
```
// the file size
$size = $sftp->size($url->path());

// fetch the import file
$result = $sftp->get($url->path());

// reset the result
$result = "";

// download the file in chunks
for ($offset = 0; $offset < $size; $offset += 1024*1024) 
{
    // download a chuck of the file
    $result .= $sftp->get($url->path(), false, $offset, 1024*1024);
}
```
Could you please mail the results to [hamza.aitmessaoud@copernica.com](hamza.aitmessaoud@copernica.com).

**Comment by terrafrost** [comment]
What version of phpseclib are you using? 1.0.14 or 2.0.14? A copy of your `phpinfo()` would be helpful as well.

There are a few things that I think, at this point, are going to significantly hinder diagnostics.

- Inability to reproduce. phpseclib passes all the unit tests and some of the unit tests involve the downloading of files. To top it off, with phpseclib being downloaded close to 1.5 million times monthly and no one else reporting the issue I think it's unlikely I'm going to have a lot of success reproducing the issue
- A review of the logs and the code reveal no obvious issues

What'd be helpful at this point (in addition to what I initially asked for) is SSH access. SSH access to the server you're downloading the files from might help but I actually kinda doubt it since the logs don't show anything to be wrong. I mean, at the very least, it couldn't hurt! But after that I think SSH access to the machine you're running phpseclib on would help.

Failing that what I guess the next approach would be...  I post debug code and you post the response. Given that there's no obvious issues in place we'll just have to double check every single step. So if I have to ask you to make 20 successive changes and it takes you 24h to get back to me with each change it'll take 20 days for us to figure out what the issue is. I can't stress to you how much of an absolute pain in the ass that is and, odds are, you'll grow weary of responding before we've figured anything out and that'll be that. SSH access is highly preferable.

**Comment by hamza-am100** [comment]
Sorry for the late response. I'm using version 1.0.14
phpinfo: [phpinfo](https://drive.google.com/open?id=1h0X4eW2nVMUyksV4FD-w2XnV7yv9-LdS)

I understand the hassle to debug this problem and i agree that providing SSH access is the best option. But the problem is that the ssh sever is not own by me. I setup my own AWS server to test the library but i couldn't reproduce the issue, the files were all correctly transferred.

What next? I'm going to ask the client to make a test environment so you can gain access to the server.

**Comment by terrafrost** [comment]
Looks like there are a number of proprietary extensions installed that I've never heard of. I could list them out but idk that I should in public lol.

Anyway, I do wonder if maybe one of those is kinda mucking things up. I Google'd for some of these seemingly proprietary extensions without success.

A test environment would def help.

Thanks!!

**Comment by hamza-am100** [comment]
Thank you for you patients.

The extensions you mentioned are developed by our self, you wont find them online. I have sent you the information for the test environment, the same problem occurs when downloading a file.

**Comment by bantu** [comment]
> The extensions you mentioned are _developed by our self_, you wont find them online.

This is a rather important detail of your setup. Please mention it in the first place next time.

> I setup my own AWS server to test the library but i couldn't reproduce the issue, the files were all correctly transferred.

Did your test environment contain these extensions? Did you try reproducing the problem with these extensions disabled? I do not know what these extensions do or how they work, but to me it seems more than likely the problem is caused by them. Would recommend excluding that this is the case.

**Comment by terrafrost** [comment]
Thanks! I'll look at it for sure this weekend but I'll try to look at it sooner if I get a chance!

**Comment by hamza-am100** [comment]
> Did your test environment contain these extensions? Did you try reproducing the problem with these extensions disabled? I do not know what these extensions do or how they work, but to me it seems more than likely the problem is caused by them. Would recommend excluding that this is actually the case.

Unfortunately that is not the cause, i got the same results when testing on a clean Ubuntu machine that never had the extensions installed.

> Thanks! I'll look at it for sure this weekend but I'll try to look at it sooner if I get a chance!

Thank your for your time.

**Comment by terrafrost** [comment]
Does https://github.com/terrafrost/phpseclib/tree/2.0 fix the issue for you?

Here's the specific commit:

https://github.com/terrafrost/phpseclib/commit/fce6063de6844904b0a3facf3ba63a17f6f7b69d

Thanks!

**Comment by hamza-am100** [comment]
>terrafrost@fce6063

That fixed the problem, the downloaded files all have the same size and the correct content:

Can the fix also be released for version 1.0?

**Comment by terrafrost** [comment]
It's in 1.0:

https://github.com/phpseclib/phpseclib/commits/1.0

Thanks for your help and patience!
